src/features/virtualDocumentTracker.ts

```diff
@@ -102,8 +143,21 @@ function logSynchronizationFailure(uri: Uri, error: any, server: OmniSharpServer
 }
 
 export default function trackVirtualDocuments(server: OmniSharpServer, eventStream: EventStream): IDisposable {
-    trackCurrentVirtualDocuments(server, eventStream);
-    let disposable = trackFutureVirtualDocuments(server, eventStream);
-    
-    return disposable;
+    let disposable: vscode.Disposable;
+    const registration = server.onProjectAdded(async () => {
+        registration.dispose();
+
+        // We wait until project state on the server has been initialized to add any virtual documents.
+
+        server.waitForEmptyEventQueue().then(async () => {
```

Comment:
@rchande @akshita31 this is ugly but essentially we need to wait until all of the projects have been initialized so we wait for the project initialization process to start and then in the background (we're not awaiting this) queue up a virtual document tracker initialization step. All of this is to ensure that we don't try and initialize a virtual document before its corresponding project is ready.

I thought about trying to assume the project in this virtual document tracker but figured that'd break functionality for anyone who truly just wanted a miscellaneous project only virtual document.

If there's a better way to do this i'm all ears.

Reply:
@NTaylorMullen Why are we trying to wait for all projects to be initialized?

Reply:
> I believe it should get "promoted" to being a non-misc file if a project loads it after it's added as a misc file. If that's not the case, we should make it so. Have you actually seen it fail to promote the file?

Seems that way, or the promotion takes a super long time. From what I looked at it seemed that only MSBuild included file are promoted once a project has been loaded. Probably missing something though. Want to sit down and person and debug?